design/Proposal-to-enhance-schema-for-metal3.md

```diff
@@ -0,0 +1,228 @@
+
+To make the existing `metal3` schema compatible and scalable for heavy computation 
+based use cases.  
+
+### Goals
+
+To compare the introspection data fetched from `Ironic-Inspector` through 
+`Gopher-cloud client` library with `metal3` schema and add useful parameters 
+in `metal3` schema.
+
+## Proposal
+
+The comparison of hardware introspection data coming from `Ironic-Inspector` 
+and existing `metal3` schema for configuration of `baremetalhost` concludes 
+that few hardware parameters are missing in `metal3` schema that are fetched 
+from `Ironic-Inspector`. We propose to add these parameters in `metal3` schema for 
+scalability and performance enhancement.
+
+### Implementation Details/Notes/Constraints
+
+Link for Existing `Metal3` Specs
+
+Please refer `metal3` spec for bare-metal:
+https://github.com/metal3-io/baremetal-operator/blob/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml
+
+The investigation of the existing schema indicates that BareMetal-Operator 
+gets BareMetal node introspection data using an API call to `Ironic-Inspector` 
+through `gopher-cloud client` library. 
+
+Having fetched BareMetal Node introspection data from Ironic-Inspector and
+compared it with BareMetalHost schema of `metal3` `(metal3.io_baremetalhosts_crd.yaml`), 
+we recommend following parameters need to be added in schema of `metal3` for `NFV` 
+workloads.
+1. Add enhancement in the existing `metal3` schema for the following parameters:
+
+    CPU Parameters:
+    
+    Parameters for `NUMA` topology:
+     
+    For better performance, selection of CPUs based on the `NUMA` topology 
+    becomes necessary. In case of nodes with `DPDK` aware NICs, the CPUs for 
+    poll mode driver (PMD) need to be selected from the `NUMA` node associated 
+    with the `DPDK` NICs. If hyperthreading is enabled, then selection of the 
+    logical cores requires the knowledge of the siblings.  
+
+    Declarative design idea:
+    ````yaml
+    Cpu
+     …
+    numa_topology:
+      properties:
+        numa_node_0:
+          properties:
+            nics:
+              properties:
+                numa_node:
+                  type: integer
+                  description: Numa node id
+                name:
+                  type: string
+                  description: network interface name
+              type: object
+            type: array
+            ram:
+              properties:
+                numa_node:
+                  type: integer
+                  description: Numa node id
+                size_kb:
+                  type: string
+                  description: memory in kb
+              type: object
+            type: array
+            cpus:
+              properties:
+                cpu:
+                  type: integer
+                  description: cpu id 
+                numa_node:
+                  type: integer
+                  description: Numa node id
+                thread_siblings:
+                  items:
```

Comment:
I don't know what thread siblings are. Is this a list of integers? What do the integers represent?

Reply:
list of integers. Integers represents the CPU's

Reply:
It seems like this information could be expressed more compactly with a list of lists of integers.

```
cpu_thread_siblings:
   - [1, 17]
   - [10, 26]
   - ...
```

How does a consumer typically use the data? Is there any reason not to favor the more compact representation?

Reply:
We can present NUMA node siblings in list of list for respective nodes. It is possible.